---
title: "Application of typical human microbiomes signatures: overlap with signatures of different diseases and exposures in the published literature."
author: "Haoyan Zhong"
date: "`r format(Sys.Date(), '%b %d, %Y')`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Applications}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Typical stool-associated signature analysis
We used control samples from the curatedMetagenomicData database to identify “health-associated, typical” genera associated with adults (ages 19 - 107) across multiple body sites, and provide these lists of signatures and prevalence in machine-readable formats for epidemiological analyses. 
This analysis aims to demonstrate the application of these signatures in high-throughput analysis- we analyze the typical stool-associated signature in the context of 3,790 stool-associated taxa of various exposures and health outcomes from 329 studies found in BugSigDB.

## Data preparation

In order to perform the analysis, we need to pull data from 2 sources:
1. Stool-associated signatures of differentially abundant genera across various health outcomes and exposures, accessed from BugSigDB.
2. typical genera identified from adult control samples in curatedMetagenomicData

```{r library, message = FALSE}
library(dplyr)
library(bugsigdbr)
library(BugSigDBStats)
library(ggpubr)
```

## Get all human stool sample studies from BugSigDB

```{r}
full.dat <-
  bugsigdbr::importBugSigDB(version = "10.5281/zenodo.6468009", cache = FALSE)
ind1 <- lengths(full.dat[["MetaPhlAn taxon names"]]) > 0
ind2 <- lengths(full.dat[["NCBI Taxonomy IDs"]]) > 0
dat <- full.dat[ind1 & ind2,] %>%
  filter (`Host species` == "Homo sapiens") %>%
  filter (`Body site` == 'feces')
```

Find the number of unique studies of human stool samples in BugSigigDB:

```{r}
length(unique(dat$Study))
```

Get genus level information from human stool sample studies:

```{r}
allbug <- bugsigdbr::getSignatures(dat, tax.level = 'genus')
```

Find the total number of total genera that were reported changed in abundance in BugSigDB human stool sample studies:

```{r}
length(unlist(allbug))
```

Find the total number of unique genera that were reported changed in abundance in BugSigDBhuman human stool sample studies:

```{r}
length(unique(unlist(allbug)))
```

## Divide signatures into 2 groups: increased and decreased relative abundance in study group

### Signatures of increased relative abundance at the genus level

```{r}
bugstudy_increased <-  dat %>%
  filter (`Abundance in Group 1` == "increased")
sig_increased <-
  bugsigdbr::getSignatures(bugstudy_increased, tax.level = 'genus')
```

The number of studies that reported signatures of increased relative abundance:

```{r}
length(unique(bugstudy_increased$Study))
```

The number of genera reported in signatures of increased relative abundance (counting duplicates):

```{r}
length(unlist(sig_increased))
```

### Signatures of _decreased_ relative abundance at the genus level

```{r}
bugstudy_decreased <-  dat %>%
  filter (`Abundance in Group 1` == "decreased")
sig_decreased <-
  bugsigdbr::getSignatures(bugstudy_decreased, tax.level = 'genus')
```

Find the number of studies that reported signature decreased in relative abundance (counting duplicates):

```{r}
length(unique(bugstudy_decreased$Study))
```

Find the number of genera that reported signature decreased in relative abundance:

```{r}
length(unlist(sig_decreased))
```


Get typical healthy signatures from GitHub:

“health-associated, typical” genera within adults identified curatedMetagenomicData database were uploaded to GitHub (https://github.com/waldronlab/TypicalMicrobiomeSignaturesExports/blob/main/matrix_genus_adult.csv). 
“health-associated, typical” genera was defined at several different prevalence thresholds- 30%, 50%, 70%

```{r}
thresh <- read.csv("https://raw.githubusercontent.com/waldronlab/TypicalMicrobiomeSignaturesExports/main/matrix_genus_adult.csv") %>%
  arrange(desc(stool_genus_prevalence))
```

### Get typical signatures at a threshold of 30%

```{r}
bugs30 <- thresh %>%
  filter (stool_genus_prevalence >=0.3) %>%
  pull("NCBI") 
```

### Get signature at 50% prevalence threshold
```{r}
bugs50 <- thresh %>%
  filter (stool_genus_prevalence >=0.5) %>%
  pull("NCBI") 
```

### Get typical signature at prevalence threshold= 70%

```{r}
bugs70 <- thresh %>%
  filter (stool_genus_prevalence >=0.7) %>%
  pull("NCBI") 
```


## Analysis 1:  

Compare the odds of overlapping with prevalent genera in healthy population using genera reported as decreased vs increased in BugSigDB study group, using thresholds of 30%, 50%, and 70% respectively


Get all genera appearing in curatedMetagenomicData:

```{r, message = FALSE}
library(curatedMetagenomicData)
library(mia)
library(TypicalMicrobiomeSignatures)
```

```{r, message = FALSE}
adult <- sampleMetadata %>%
  filter(age_category %in% (c("adult", "senior"))) %>%
  filter(body_site == "stool") %>%
  filter(study_name != "FengQ_2015") %>%
  filter(study_name != "YuJ_2015")
adult <- adult[, colSums(is.na(adult)) < nrow(adult)]

cmg <- calcPrevalence(adult, rank = "genus")
```

Now combine all genus appeared in curatedMetagenomicData and BugSigDB:

```{r}
allbuglist <- array(unlist(allbug))
full <- cmg %>%
  select ("NCBI") %>%
  rbind(allbuglist) %>%
  distinct(NCBI)
length(full$NCBI)
```

Report odds (decresed/incresed) of overlapping with prevalent genera in healthy population:

```{r} 
defull <- data.frame(unlist(sig_decreased)) %>%
  mutate(bugs30 = ifelse(unlist.sig_decreased. %in% bugs30, 1, 0)) %>%
  mutate(bugs50 = ifelse(unlist.sig_decreased. %in% bugs50, 1, 0)) %>%
  mutate(healthy70 = ifelse(unlist.sig_decreased. %in% bugs70, 1, 0)) %>%
  mutate(decrease = 1) %>%
  dplyr::rename(NCBI= 'unlist.sig_decreased.')

infull <- data.frame(unlist(sig_increased)) %>%
  mutate(bugs30 = ifelse(unlist.sig_increased. %in% bugs30, 1, 0)) %>%
  mutate(bugs50 = ifelse(unlist.sig_increased. %in% bugs50, 1, 0)) %>%
  mutate(bugs70 = ifelse(unlist.sig_increased. %in% bugs70, 1, 0)) %>%
  mutate(decrease = 0) %>%
  dplyr::rename(NCBI='unlist.sig_increased.')

fullf <- rbind(defull, infull)

m30 <- table(x = fullf$decrease, y = fullf$bugs30)
m50 <- table(x = fullf$decrease, y = fullf$bugs50)
m70 <- table(x = fullf$decrease, y = fullf$bugs70)

library(epitools)
oddsratio(m30)
oddsratio(m50)
oddsratio(m70)
```

## Analysis 2

Assess the association between the change in relative abundance in study groups in literature (decreased or increased) and the prevalence of typical genera found in healthy volunteers’ stool samples.

First, get the prevalence of each typical health-associated genera from curatedMetagenomicData healthy volunteers’ stool samples.

```{r, message = FALSE}
healthystool <- sampleMetadata %>%
  filter(age_category %in% (c("adult", "senior"))) %>%
  filter (disease=="healthy") %>% 
  filter(body_site == "stool") %>%
  filter(study_name != "FengQ_2015") %>%
  filter(study_name != "YuJ_2015")

healthystool <- healthystool[,colSums(is.na(healthystool))<nrow(healthystool)]
healthyp<-calcPrevalence(adult, rank = "genus")
```

Calculate the frequency (overall, increased, decreased) of each genera reported in BugSigDB studies.
```{r}
fullf1 <- fullf %>%
  mutate(x = 1)
tbl <- with(fullf1, table(NCBI, x))
tbld <- with(fullf1, table(NCBI, decrease))

tbl1 <- as.data.frame(tbl)
names(tbl1)[3] <- 'total'

tbld1 <- as.data.frame(tbld)
tblm <- tbl1 %>%
  left_join(tbld1, by = "NCBI") %>%
  mutate(NBI = as.factor(NCBI))
```

Now combine typical health-associated genera from curatedMetagenomicData healthy volunteers’ stool samples with signature from BugSigDB human tool sample that changed in abundance. 

```{r}
healthypp <- healthyp %>%
  mutate(NCBI = trimws(NCBI)) %>%
  mutate(NCBI = as.factor(NCBI))

tbl2 <- tblm %>%
  left_join(healthypp, by = "NCBI")

levels(tbl2$decrease) <- c("Increased", "Decreased")
tbl2[is.na(tbl2)] <- 0 
```

Include only genera with prevalence of at least 5% in curatedMetagenomicData:

```{r}
tbl3<- tbl2 %>%
  filter (value>0.05)
```

### Linear regression

Calculate the frequency of genera mentioned in BugSigDB~ prevelance of genera in healthy control samples * increased/decreased in BugSigDB:

```{r}
l<-lm(Freq ~ value * decrease, data=tbl3)
summary(l)
```
### Figure 1. 

Correlation between the prevalence of typical genera in the stool of control subjects of curatedMetagenomicData (including only genera with prevalence of at least 5%) and the frequency with which the genera are reported among signatures of decreased or increased relative abundance in study groups of bugsigdb.org.

```{r}
d <- tbl3 %>% filter (decrease == 'Decreased')
i <- tbl3 %>% filter (decrease == 'Increased')

p1 <- ggplot(data = d, aes(x = value * 100, y = Freq)) +
  geom_point(size = 3, shape = 6) +
  xlab("Prevalence of genera among healthy individuals %") +
  ylab("Count of genera decreased in relative abundance reported in BugSigDB") +
  geom_smooth(method = "lm", color = 'black') +
  ylim(0, 90)

p2 <- ggplot(data = i, aes(x = value * 100, y = Freq)) +
  geom_point(size = 3, shape = 17) +
  xlab("Prevalence of genera among healthy individuals %") +
  ylab("Count of genera increased in relative abundance reported in BugSigDB") +
  geom_smooth(method = "lm", color = 'black') +
  ylim(0, 90)

library(gridExtra)
grid.arrange(p2, p1, nrow = 1)
```

### Logistic regression model

A logistic regression model was used to assess the association between the change in relative abundance in study groups in literature (decreased or increased) with the prevalence of typical genera found in healthy volunteers’ stool samples. Odds ratio (decreased/increased) and 95% CIs are reported. 

```{r}
m <-
  glm(
    decrease ~ value,
    data = tbl3,
    weights = Freq,
    family = binomial("logit")
  )
summary(m)


exp(unname(m$coefficients[2]) * 0.1)
exp(unname(m$coefficients[2]) * 0.1 - 1.96 * unname(sqrt(diag(vcov(m)))[2]*0.1))
exp(unname(m$coefficients[2]) * 0.1 + 1.96 * unname(sqrt(diag(vcov(m)))[2]*0.1))
```
