---
title: "BugSigDBStats"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{BugSigDBStats}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## BugsigDB stat- Signature overlaps

Compare genus changed in abundance from all diseased subjects in BugsigDB to existing typical genus from healthy volunteers recorded in BugsigDB Study 562
typical genus were set at prevalence level=30, 50, and 70

## Get typical healthy signatures from Study 562
Using feces sample as an example
```{r library}
library(dplyr)
library(bugsigdbr)
library(BugSigDBStats)
library(ggpubr)
```

```{r BugsigDB}
full.dat <- bugsigdbr::importBugSigDB(version = "10.5281/zenodo.6468009", cache=FALSE) 

#Stripping empty signatures:
ind1 <- lengths(full.dat[["MetaPhlAn taxon names"]]) > 0
ind2 <- lengths(full.dat[["NCBI Taxonomy IDs"]]) > 0
dat <- full.dat[ind1 & ind2,] %>%
  filter (`Host species`=="Homo sapiens") %>%
  filter (`Body site` == 'feces')

#find number of unique human feces studies in BugSigigDB
length(unique(dat$Study))
```

```{r BugSigDB genera}
allbug <- bugsigdbr::getSignatures(dat, tax.level='genus')
NCBI <- array(unlist(allbug))
allbuglist <- data.frame(NCBI) %>%
  distinct(NCBI)
#total number of unique genera that were reported changed in abundance in BugSigDB
length(unique(allbuglist$NCBI))
```

```{r healthy30}
#prevalence threshold=30
bugs30<- (dat) %>%
  filter (Study=='Study 562') %>%
  filter (Experiment=="Experiment 11") %>%
  filter (`Body site` == "feces") %>%
  filter (grepl("genus",Description))
healthysig30 <- bugsigdbr::getSignatures(bugs30)  
head(healthysig30)
```

```{r healthy50}
#prevalence threshold=50%
bugs50<- (dat) %>%
  filter (Study=='Study 562') %>%
  filter (Experiment=="Experiment 3" & `Signature page name`=='Signature 1') %>%
  filter (`Body site` == "feces") %>%
  filter (grepl("genus",Description))
healthysig50 <- bugsigdbr::getSignatures(bugs50)  
head(healthysig50)
u50<-unlist(healthysig50)
```

```{r healthy70}
#prevalence threshold=70%
bugs70<- (dat) %>%
  filter (Study=='Study 562') %>%
  filter (Experiment=="Experiment 9" & `Signature page name`=='Signature 1') %>%
  filter (`Body site` == "feces") %>%
  filter (grepl("genus",Description))
healthysig70 <- bugsigdbr::getSignatures(bugs70)  
head(healthysig70)

#hardcode- change error in BugSigDB, will remove after next release
u70<-unlist(healthysig70)
u70[22]='1905344'

```
## Get disease signature

```{r increased_disease_genus}
bugdisease_increased <-  dat %>%
  filter (`Abundance in Group 1`=="increased") 
length(unique(bugdisease_increased$Study))
#Get genus level changed in abundance studies only
diseasesig_increased <- bugsigdbr::getSignatures(bugdisease_increased, tax.level='genus')
length(diseasesig_increased)

NCBI_increased <- array(unlist(diseasesig_increased))
#all reported increased genera from BugsigDB
increasedist <- data.frame(NCBI_increased) 
distinct_i<-unique(increasedist)
count(distinct_i)
count(increasedist) 
#A total of 299 studies reported - 232 distinct increased genera, #total 1676 genera
```

```{r decreased_disease_genus}
bugdisease_decreased <-  dat %>%
  filter (`Abundance in Group 1`=="decreased") 
length(unique(bugdisease_decreased$Study))
#get genus level sig change in abundance only
diseasesig_decreased <- bugsigdbr::getSignatures(bugdisease_decreased, tax.level='genus')
length(diseasesig_decreased)

NCBI_decreased <- array(unlist(diseasesig_decreased))
#all reported increased genera from BugsigDB
decreasedist <- data.frame(NCBI_decreased)
distinct_d<-unique(decreasedist)
count(distinct_d)
count(decreasedist) 
#A total of 308 studies reported - 243 distinct decreased genera, #total 1700 genera
```

```{r all cmg stool NCBI}
library(curatedMetagenomicData)
library(dplyr)
library(mia)
#keep adult only
adult <- sampleMetadata %>%
  filter(age_category %in% (c("adult", "senior"))) 
#drop empty entry
adult <- adult[,colSums(is.na(adult))<nrow(adult)]
#select stool
stool <- adult %>% filter(body_site == "stool")
df_sub1 <-
  stool %>% returnSamples("relative_abundance", rownames = "NCBI")
df_sub2 <-
  stool %>% returnSamples("relative_abundance", rownames = "short")

prev <- function(prevalencecutoff = 0.000001) {
  .returnSig <- function(obj, threshold = 0, prevalence=0.000001){
    fractionpassing.logical <- rowSums(obj > threshold) / ncol(obj)
    rows.passing <- fractionpassing.logical > prevalence
    return(fractionpassing.logical[rows.passing])
  }

  df_sub.byranks <- mia::splitByRanks(df_sub1)
  df_subg1 <-
    .returnSig(assay(df_sub.byranks[["genus"]]), prevalence = prevalencecutoff)
  df_subs1 <-
    .returnSig(assay(df_sub.byranks[["species"]]), prevalence = prevalencecutoff)
  
  df_sub.byranks <- mia::splitByRanks(df_sub2)
  df_subg2 <-
    .returnSig(assay(df_sub.byranks[["genus"]]), prevalence = prevalencecutoff)
  df_subs2 <-
    .returnSig(assay(df_sub.byranks[["species"]]), prevalence = prevalencecutoff)
  
  df_subg2 <- tibble(name = names(df_subg2), value = df_subg2)
  df_subs2 <- tibble(name = names(df_subs2), value = df_subs2)
  df_sub_genus <- data.frame(df_subg1, df_subg2) %>%
    arrange(desc(value))
}
stoollist<-prev(prevalencecutoff = 0.000001)
stoollist1 <- tibble::rownames_to_column(stoollist, "NCBI")
write.csv(stoollist1,"/mnt/STORE1/bighome/haoyanzh/DE/allstool_genus.csv", row.names = FALSE)
```

```{r universe}
library(readr)
#get universe - genus that appeared in healthy and unhealthy subjects stool sample from CuratedMetagenomicData
#appeared in at least 5% of the samples - reduce false positive, N=110
cmg <- read_csv("/mnt/STORE1/bighome/haoyanzh/DE/allstool_genus.csv",show_col_types = FALSE) %>%
#  filter (value>=0.05) %>%
  select (NCBI)
length(cmg$NCBI)
#get all detectable genera in bugsigdb
full<-rbind(allbuglist,cmg) %>%
  distinct(NCBI)
length(full$NCBI) #581- #489 (from any BugSigDB study) +  441 (cmd)
```

```{r enrichment}
DE_d=as.integer(distinct_d$NCBI_decreased)
DE_i=as.integer(distinct_i$NCBI_increased)

DE_dfull=as.integer(decreasedist$NCBI_decreased)
DE_ifull=as.integer(increasedist$NCBI_increased)

healthy30=as.integer(unlist(healthysig30))
healthy50=as.integer(u50)
healthy70=as.integer(u70)
```

# Method 2 - unit is each Bugsig DB reported genus (changed in abundance)
```{r overlap pct 30 - do not take out duplicate BugsigDB genera, unit of analysis is each reported changed genus}
defull<-decreasedist %>%
  mutate(healthy30 = ifelse(NCBI_decreased %in% healthy30, 1, 0)) %>%
  mutate(healthy50 = ifelse(NCBI_decreased %in% healthy50, 1, 0)) %>%
  mutate(healthy70 = ifelse(NCBI_decreased %in% healthy70, 1, 0)) %>%
  mutate(decrease=1) %>%
  rename(NCBI=NCBI_decreased)

infull<-increasedist%>%
  mutate(healthy30 = ifelse(NCBI_increased %in% healthy30, 1, 0)) %>%
  mutate(healthy50 = ifelse(NCBI_increased %in% healthy50, 1, 0)) %>%
  mutate(healthy70 = ifelse(NCBI_increased %in% healthy70, 1, 0)) %>%
  mutate(decrease=0) %>%
  rename(NCBI=NCBI_increased)

fullf<-rbind(defull,infull)

table(fullf$decrease,fullf$healthy30)
table(fullf$decrease,fullf$healthy50)
table(fullf$decrease,fullf$healthy70)
```

```{r comparison}
m30<-table(x=fullf$decrease,y=fullf$healthy30)
m50<-table(x=fullf$decrease,y=fullf$healthy50)
m70<-table(x=fullf$decrease,y=fullf$healthy70)

library(epitools)
oddsratio(m30)
oddsratio(m50)
oddsratio(m70)
```

```{r all healthy stool NCBI}
library(curatedMetagenomicData)
library(dplyr)
library(mia)
#keep adult only
adult <- sampleMetadata %>%
  filter(age_category %in% (c("adult", "senior"))) %>%
  filter (disease=="healthy")
#drop empty entry
adult <- adult[,colSums(is.na(adult))<nrow(adult)]
#select stool
stool <- adult %>% filter(body_site == "stool")
df_sub1 <-
  stool %>% returnSamples("relative_abundance", rownames = "NCBI")
df_sub2 <-
  stool %>% returnSamples("relative_abundance", rownames = "short")

prev <- function(prevalencecutoff = 0.000001) {
  .returnSig <- function(obj, threshold = 0, prevalence=0.000001){
    fractionpassing.logical <- rowSums(obj > threshold) / ncol(obj)
    rows.passing <- fractionpassing.logical > prevalence
    return(fractionpassing.logical[rows.passing])
  }

  df_sub.byranks <- mia::splitByRanks(df_sub1)
  df_subg1 <-
    .returnSig(assay(df_sub.byranks[["genus"]]), prevalence = prevalencecutoff)
  df_subs1 <-
    .returnSig(assay(df_sub.byranks[["species"]]), prevalence = prevalencecutoff)
  
  df_sub.byranks <- mia::splitByRanks(df_sub2)
  df_subg2 <-
    .returnSig(assay(df_sub.byranks[["genus"]]), prevalence = prevalencecutoff)
  df_subs2 <-
    .returnSig(assay(df_sub.byranks[["species"]]), prevalence = prevalencecutoff)
  
  df_subg2 <- tibble(name = names(df_subg2), value = df_subg2)
  df_subs2 <- tibble(name = names(df_subs2), value = df_subs2)
  df_sub_genus <- data.frame(df_subg1, df_subg2) %>%
    arrange(desc(value))
}
stoollist<-prev(prevalencecutoff = 0.000001)
stoollist2 <- tibble::rownames_to_column(stoollist, "NCBI")
write.csv(stoollist2,"/mnt/STORE1/bighome/haoyanzh/DE/healthystool_genus.csv", row.names = FALSE)
```

```{r interection}
fullf1<-fullf %>%
  mutate(x=1)

tbl <- with(fullf1, table(NCBI, x))
tbld <- with(fullf1, table(NCBI, decrease))

tbl1<- as.data.frame(tbl)
names(tbl1)[3] <- 'total'

tbld1<-as.data.frame(tbld) 
tblm<-tbl1 %>%
  left_join(tbld1, by="NCBI")

tblm$NCBI <- as.factor(tblm$NCBI) 

healthyp <- read_csv("/mnt/STORE1/bighome/haoyanzh/DE/healthystool_genus.csv",show_col_types = FALSE) 
healthyp$NCBI <- as.factor(healthyp$NCBI)   
tbl2<- tblm %>%
  left_join(healthyp, by="NCBI")
levels(tbl2$decrease) <- c("Increased", "Decreased")
tbl2[is.na(tbl2)] <- 0 

l<-lm(Freq ~ value * decrease, data=tbl2)
summary(l)

d<-tbl2 %>% filter (decrease=='Decreased') 
i<-tbl2 %>% filter (decrease=='Increased') 

p1<-ggplot(data=d, aes(x = value*100, y = Freq)) +
  geom_point(size=3, shape=6) +
  xlab("Prevalence of genus among healthy individuals") +
  ylab("# decreased genus reported in BugSigDB") + 
  geom_smooth(method = "lm", color='black') + 
  ylim(0, 87)

p2<- ggplot(data=i, aes(x = value*100, y = Freq)) +
  geom_point(size=3, shape=17) +
  xlab("Prevalence of genus among healthy individuals") +
  ylab("# increased genus reported in BugSigDB") + 
  geom_smooth(method = "lm", color='black') + 
  ylim(0, 87)

library(gridExtra) 
grid.arrange(p1, p2, nrow = 1)
```

```{r corr between #genus reported changed in BugSigDB and prevelance in healthy cmd subjects using healthy genus threshold=50%}
tbl3<- tbl2 %>%
  filter (value>0.05)

l<-lm(Freq ~ value * decrease, data=tbl3)
summary(l)

ggplot(as.data.frame(tbl3), aes(x = value*100, y = Freq, shape = decrease)) +
  geom_point(size=3)+
  scale_shape_manual(values=c(17,6)) +
  xlab("Prevalence of genus among healthy individuals") +
  ylab("# signatures genus appear in BugSigDB") + 
  guides(color = guide_legend(title = "Changed in abundance")) +
  geom_smooth(method = "lm")
 
d<-tbl3 %>% filter (decrease=='Decreased') 
i<-tbl3 %>% filter (decrease=='Increased') 

p1<-ggplot(data=d, aes(x = value*100, y = Freq)) +
  geom_point(size=3, shape=6) +
  xlab("Prevalence of genus among healthy individuals") +
  ylab("# decreased genus reported in BugSigDB") + 
  geom_smooth(method = "lm", color='black') + 
  ylim(0, 87)

p2<- ggplot(data=i, aes(x = value*100, y = Freq)) +
  geom_point(size=3, shape=17) +
  xlab("Prevalence of genus among healthy individuals") +
  ylab("# increased genus reported in BugSigDB") + 
  geom_smooth(method = "lm", color='black') + 
  ylim(0, 87)

library(gridExtra) 
grid.arrange(p1, p2, nrow = 1)
```
